swarm/network/simulations/discovery/dynamic_test.go

```diff
@@ -0,0 +1,359 @@
+			var j int
+			if i < bootNodeCount {
+				if i == bootNodeCount-1 {
+					j = 0
+				} else {
+					j = i + 1
+				}
+			} else {
+				j = rand.Intn(bootNodeCount)
+			}
+			go func(i, j int) {
+				net.Connect(ids[i], ids[j])
+				// TODO: replace with simevents check
+				time.Sleep(time.Second * 2)
+				trigger <- ids[i]
+			}(i, j)
+		}
+		return nil
+	}
+
+	// construct the peer pot, so that kademlia health can be checked
+	check := func(ctx context.Context, id discover.NodeID) (bool, error) {
+		log.Warn("checking", "id", id)
+		var err error
+		select {
+		case <-ctx.Done():
+			err = ctx.Err()
+		case e := <-events:
+			if e.Type == p2p.PeerEventTypeAdd {
+				log.Info("check got add", "id", id)
+				n := net.GetNode(id)
+				err := ctrl.checkHealth(n)
+				log.Debug("check health return", "id", id, "err", err)
+				return true, err
+			}
+		case err := <-subs[id].Err():
+			if err != nil {
+				log.Error(fmt.Sprintf("error getting peer events for node %v", id), "err", err)
+			}
+		default:
+		}
+		return false, err
+	}
+
+	timeout := 30 * time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	result := simulations.NewSimulation(net).Run(ctx, &simulations.Step{
+		Action:  action,
+		Trigger: trigger,
+		Expect: &simulations.Expectation{
+			Nodes: ids,
+			Check: check,
+		},
+	})
+	if result.Error != nil {
+		t.Fatal(result.Error)
+	}
+
+	wg := sync.WaitGroup{}
+	for i := 0; i < int(numUpDowns); i++ {
+		wg.Add(1)
+		dur := ctrl.randomDelay(0)
+		time.Sleep(dur)
+		go func() {
+			if err := ctrl.nodeUpDown(); err != nil {
+				wg.Done()
+				t.Fatal(err)
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+
+	log.Warn("exiting test")
+}
+
+type nodeCtrl struct {
+	seq                int
+	net                *simulations.Network
+	upNodes            []discover.NodeID
+	upAddrs            [][]byte
+	nodeCursor         int
+	pot                map[discover.NodeID]*network.PeerPot
+	mu                 sync.Mutex
+	healthCheckRetries int
+	healthCheckDelay   time.Duration
+}
+
+func newNodeCtrl(net *simulations.Network, nodes []discover.NodeID) *nodeCtrl {
+	ctrl := &nodeCtrl{
+		net:                net,
+		healthCheckRetries: defaultHealthCheckRetires,
+		healthCheckDelay:   defaultHealthCheckDelay,
+	}
+	for ctrl.nodeCursor = 0; ctrl.nodeCursor < len(nodes); ctrl.nodeCursor++ {
+		ctrl.upNodes = append(ctrl.upNodes, nodes[ctrl.nodeCursor])
+		ctrl.upAddrs = append(ctrl.upAddrs, network.ToOverlayAddr(ids[ctrl.nodeCursor].Bytes()))
+		log.Debug("init nodeCtrl", "idx", ctrl.nodeCursor, "upnode", ctrl.upNodes[ctrl.nodeCursor], "upaddr", fmt.Sprintf("%x", ctrl.upAddrs[ctrl.nodeCursor]))
+	}
+	ctrl.pot = network.NewPeerPot(testMinProxBinSize, ctrl.upNodes, ctrl.upAddrs)
+	for k, p := range ctrl.pot {
+		for i, nn := range p.NNSet {
+			log.Debug("init nodeCtrl nnset", "i", i, "node", k.TerminalString(), "nn", fmt.Sprintf("%x", nn))
+		}
+	}
+	return ctrl
+}
+
+func (self *nodeCtrl) nodeUpDown() error {
```

Comment:
A short description of the strategy helps understanding the dynamics without having to read the code first...

Reply:
Indeed, but so far it's still in progress...:)